Google.Cloud.EntityFrameworkCore.Spanner.Tests/EntityFrameworkSessionLeakMockServerTests.cs

```diff
@@ -0,0 +1,2286 @@
+            optionsBuilder
+                .UseSpanner(new SpannerConnection(builder), _ => SpannerModelValidationConnectionProvider.Instance.EnableDatabaseModelValidation(false))
+                .UseMutations(MutationUsage.Never)
+                .UseLazyLoadingProxies();
+        }
+    }
+
+    /// <summary>
+    /// Tests CRUD operations using an in-mem Spanner mock server.
+    /// </summary>
+    public class EntityFrameworkSessionLeakMockServerTests : IClassFixture<SpannerMockServerFixture>
+    {
+        private readonly SpannerMockServerFixture _fixture;
+        private readonly SessionPoolManager _manager;
+
+        public EntityFrameworkSessionLeakMockServerTests(SpannerMockServerFixture service)
+        {
+            _fixture = service;
+            service.SpannerMock.Reset();
+            
+            var options = new SessionPoolOptions
+            {
+                MinimumPooledSessions = 1,
+                MaximumActiveSessions = 2,
+                WaitOnResourcesExhausted = ResourcesExhaustedBehavior.Fail
+            };
+            _manager = SessionPoolManager.Create(options);
+        }
+
+        private string ConnectionString => $"Data Source=projects/p1/instances/i1/databases/d1;Host={_fixture.Host};Port={_fixture.Port}";
+        
+        private static async Task Repeat(int count, Func<Task> action)
+        {
+            for (var i = 0; i < count; i++)
+            {
+                await action();
+            }
+        }
+
+        private async Task Repeat(Func<Task> action)
+        {
+            await Repeat(_manager.SessionPoolOptions.MaximumActiveSessions + 1, action);
+        }
+
+        /// <summary>
+        /// Creates a Entity Framework Database Context with a limited number of sessions and
+        /// that is configured to fail if all sessions in the pool have been exhausted.
+        /// </summary>
+        /// <returns>A DbContext with a limited number of sessions</returns>
+        private LimitedSessionsSampleDbContext CreateContext() => new (ConnectionString, _manager);
+        
+        [Fact]
+        public async Task FindSingerAsync_ReturnsNull_IfNotFound()
+        {
+            var sql = $"SELECT `s`.`SingerId`, `s`.`BirthDate`, `s`.`FirstName`, `s`.`FullName`, `s`.`LastName`, " +
+                $"`s`.`Picture`{Environment.NewLine}FROM `Singers` AS `s`{Environment.NewLine}" +
+                $"WHERE `s`.`SingerId` = @__p_0{Environment.NewLine}LIMIT 1";
+            _fixture.SpannerMock.AddOrUpdateStatementResult(sql, StatementResult.CreateResultSet(
+                new List<Tuple<V1.Type, string>> { },
+                new List<object[]> { }
+            ));
+
+            using var db = CreateContext();
+            await Repeat(async () =>
+            {
+                var singer = await db.Singers.FindAsync(1L);
+                Assert.Null(singer);
+            });
+        }
+
+        [Fact]
+        public async Task FindSingerAsync_ReturnsInstance_IfFound()
```

Comment:
Sorry, I should have clarified the context a bit more here. There was some concern that the Entity Framework provider itself was leaking sessions when being used normally. I therefore copy-pasted a large set of existing test cases into this file to verify that that is not the case. If any of these tests would cause a session leak (now or in the future), the test will fail, because:
1. The `LimitedSessionsSampleDbContext` that is used by all these tests have `MaxSessions=2` and `ResourcesExhaustedBehavior.Fail`
2. The tests are repeated 3 times (the `await Repeat` call below)
3. If the test case itself would leak a session, then the third attempt would fail with a `ResourceExhausted` error.

Reply:
Thanks for the clarification.